src/typ/environment.rs

```diff
@@ -467,21 +482,113 @@ impl<'a, 'b> Environment<'a, 'b> {
+                    })
+                }
+                (name, (ValueKind::ImportedTypeAndConstructor, location, 0)) => {
+                    self.warnings.push(Warning::UnusedType {
+                        name,
+                        imported: true,
+                        location,
+                    })
+                }
+                (name, (ValueKind::PrivateConstant, location, 0)) => self
+                    .warnings
+                    .push(Warning::UnusedPrivateModuleConstant { name, location }),
+                (name, (ValueKind::PrivateConstructor(_), location, 0)) => {
+                    self.warnings.push(Warning::UnusedConstructor {
+                        name,
+                        imported: false,
+                        location,
+                    })
+                }
+                (name, (ValueKind::PrivateFunction, location, 0)) => self
+                    .warnings
+                    .push(Warning::UnusedPrivateFunction { name, location }),
+                (name, (ValueKind::PrivateType, location, 0)) => {
+                    self.warnings.push(Warning::UnusedType {
+                        name,
+                        imported: false,
+                        location,
+                    })
+                }
+                (name, (ValueKind::ImportedValue, location, 0)) => self
+                    .warnings
+                    .push(Warning::UnusedImportedValue { name, location }),
+
+                (name, (ValueKind::Variable, location, 0)) => self
+                    .warnings
+                    .push(Warning::UnusedVariable { name, location }),
+
+                _ => {}
+            };
         }
-        for (name, location) in self.unused_private_mixed_constructors.drain() {
-            self.warnings
-                .push(Warning::UnusedConstructor { name, location })
+    }
+
+    pub fn init_value_usage(&mut self, name: String, kind: ValueKind, location: SrcSpan) {
+        match self
+            .value_usages
+            .front_mut()
+            .unwrap()
+            .insert(name.to_string(), (kind, location, 0))
+        {
+            Some((kind, location, 0)) => {
+                // a value was overwritten in the top most scope without being used
+                let mut unused = HashMap::with_capacity(1);
+                let _ = unused.insert(name, (kind, location, 0));
+                self.handle_unused_local_values(unused);
+            }
+            _ => {}
         }
     }
 
     pub fn type_used(&mut self, name: &str) {
-        let _ = self.unused_private_mixed_constructors.remove(name);
-        let _ = self.unused_private_types.remove(name);
+        self.increment_usage(name);
     }
 
     pub fn value_used(&mut self, name: &str) {
-        let _ = self.unused_private_mixed_constructors.remove(name);
+        self.increment_usage(name);
+    }
+
+    fn increment_usage(&mut self, name: &str) {
+        let mut val_found = false;
```

Comment:
In this function we can use `break` to stop iteration when we find the value rather than continuing but skipping the body with an `if`